xmrstak/misc/executor.cpp

```diff
@@ -1241,6 +1259,111 @@ void executor::http_json_report(std::string& out)
 	out = std::string(bigbuf.get(), bigbuf.get() + bb_len);
 }
 
+void executor::http_prometheus_report(std::string& out)
+{
+#ifndef CONF_NO_PROMETHEUS
+	// Set Hash rates
+	double fTotal[3] = { 0.0, 0.0, 0.0};
+	size_t nthd = pvThreads->size();
+
+	for(size_t i=0; i < nthd; i++) {
+		double fHps[3];
+
+		fHps[0] = telem->calc_telemetry_data(10000, i);
+		fHps[1] = telem->calc_telemetry_data(60000, i);
+		fHps[2] = telem->calc_telemetry_data(900000, i);
+
+		sMetric.vgHashRates[i * 3]->Set(fHps[0]);
+		sMetric.vgHashRates[i * 3 + 1]->Set(fHps[1]);
+		sMetric.vgHashRates[i * 3 + 2]->Set(fHps[2]);
+
+		fTotal[0] += fHps[0];
+		fTotal[1] += fHps[1];
+		fTotal[2] += fHps[2];
+	}
+	sMetric.agHashRateTotal[0]->Set(fTotal[0]);
+	sMetric.agHashRateTotal[1]->Set(fTotal[1]);
+	sMetric.agHashRateTotal[2]->Set(fTotal[2]);
+
+	sMetric.gPoolPing->Set(42);
+
+	auto& registry = xmrstak::environment::inst().pRegistry;
+
+	// Serialize for output.
+	auto serializer = std::unique_ptr<prometheus::Serializer>{};
+	serializer.reset(new prometheus::TextSerializer());
+
+	out = serializer->Serialize(registry->get()->Collect());
+#else
+	out = std::string("This build has not enabled Prometheus metrics.\nPlease rebuild with -DPROMETHEUS_ENABLE=ON option.");
+#endif
+}
+
+#ifndef CONF_NO_PROMETHEUS
+void executor::init_metrics() {
+	auto& env = xmrstak::environment::inst();
+
+	using namespace prometheus;
+	auto& counter_family_results = BuildCounter()
+		.Name("xmrstak_submitted_results_total")
+		.Register(**env.pRegistry);
+	sMetric.cSubmittedResultsAccepted = &counter_family_results.Add(
+		{{"type", "accepted"}}
+	);
+	sMetric.cSubmittedResultsRejected = &counter_family_results.Add(
+		{{"type", "rejected"}}
+	);
+
+	auto& gauge_family_diff = BuildGauge()
+		.Name("xmrstak_difficulty_bucket")
+		.Register(**env.pRegistry);
+	sMetric.gDifficulty = &gauge_family_diff.Add({});
+
+	auto& gauge_family_pool_ping = BuildGauge()
+		.Name("xmrstak_pool_connection_ping_seconds")
+		.Register(**env.pRegistry);
+    // TODO: Implement fetching pool address.
+	sMetric.gPoolPing = &gauge_family_pool_ping.Add({{"address", "xxx:3333"}});
+
+	auto& gauge_family_hr_total = BuildGauge()
+		.Name("xmrstak_hash_rate_bytes_total")
+		.Register(**env.pRegistry);
+	sMetric.agHashRateTotal[0] = &gauge_family_hr_total.Add({{"le", "10"}});
+	sMetric.agHashRateTotal[1] = &gauge_family_hr_total.Add({{"le", "60"}});
+	sMetric.agHashRateTotal[2] = &gauge_family_hr_total.Add({{"le", "900"}});
+
+	auto& gauge_family_hr = BuildGauge()
+		.Name("xmrstak_hash_rate_bytes")
```

Comment:
I would remove _bytes from here the metric name. Also it would be nice to have device and thread_id labels here too.

Reply:
Bytes represent units: https://prometheus.io/docs/practices/naming/. Labels already exits, look at 1348-1362 there are dynamically created.

Reply:
Missed that. Thanks!